x/mongo/driver/operation/hello_test.go

```diff
@@ -0,0 +1,874 @@
+package operation
+
+import (
+	"runtime"
+	"testing"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/internal/assert"
+	"go.mongodb.org/mongo-driver/internal/require"
+	"go.mongodb.org/mongo-driver/version"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
+const documentSize = 5         // 5 bytes to start and end a document
+const embeddedDocumentSize = 7 // 7 bytes to append a document element
+const stringElementSize = 7    // 7 bytes to append a string element
+const int32ElementSize = 6     // 6 bytes to append an int32 element
+
+func assertAppendClientMaxLen(t *testing.T, got bsoncore.Document, wantD bson.D, maxLen int) {
+	t.Helper()
+
+	tooLarge := len(got)-documentSize > maxLen
+	require.False(t, tooLarge, "got document is too large: %v", got)
+
+	var gotD bson.D
+	err := bson.Unmarshal(got, &gotD)
+	require.NoError(t, err, "error unmarshaling got document: %v", err)
+
+	assert.Equal(t, wantD, gotD, "got %v, want %v", gotD, wantD)
+}
+
+func encodeWithCallback(t *testing.T, cb func(int, []byte) ([]byte, error)) bsoncore.Document {
+	t.Helper()
+
+	var err error
+	idx, dst := bsoncore.AppendDocumentStart(nil)
+
+	dst, err = cb(len(dst), dst)
+	require.NoError(t, err, "error appending client metadata: %v", err)
+
+	dst, err = bsoncore.AppendDocumentEnd(dst, idx)
+	require.NoError(t, err, "error appending document end: %v", err)
+
+	got, _, ok := bsoncore.ReadDocument(dst)
+	require.True(t, ok, "error reading document: %v", got)
+
+	return got
+}
+
+func addMaxLenStringElem(key, name string) func() int {
+	return func() int {
+		return stringElementSize + len(key) + len(name)
+	}
+}
+
+//nolint:unparam
+func addMaxLenInt32Elem(key string) func() int {
+	return func() int {
+		return int32ElementSize + len(key)
+	}
+}
+
+func addMaxLenBuf(subtract int) func() int {
+	return func() int {
+		return subtract
+	}
+}
+
+func addMaxLenEmbeddedDocument(key string) func() int {
+	return func() int {
+		return embeddedDocumentSize + len(key)
+	}
+}
+
+func calcMaxLen(fn ...func() int) int {
```

Comment:
Optional: Expressing the integer addition logic as a series of functions is unnecessarily difficult to understand compared to using arithmetic operators. Consider removing the additional function layer and add the calculated sizes directly.

For example, consider the following calculation:
```go
calcMaxLen(
	addMaxLenEmbeddedDocument("driver"),
	addMaxLenStringElem("name", driverName),
	addMaxLenStringElem("version", version.Driver),
	addMaxLenBuf(1)),
)
```

That can be rewritten as a simple arithmetic operation:
```go
lenEmbeddedDocument("driver") +
	lenStringElem("name", driverName) +
	lenStringElem("version", version.Driver) +
	-1
```

The latter is more concise and doesn't require the reader to understand that `calcMaxLen` adds the lengths together or that `addMaxLenBuf` adds the given number from the calculated length.

Reply:
Good idea, I've updated the code removing all of the helper functions doing byte math. Favoring building the document through the bsoncore package, per [this](https://github.com/mongodb/mongo-go-driver/pull/1212#discussion_r1160400905) comment.